File: cpp/day11.cpp
#include <fstream>
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <map>
#include <utility>
#include <queue>

std::vector<std::string> split(std::string&& s, char c) {
	std::vector<std::string> temp;
	while(true) {
		size_t end = s.find(c);
		if(end == std::string::npos) {
			temp.push_back(s);
			break;
		} else {
			temp.push_back(s.substr(0,end));
		}
		s = s.substr(end + 1);
	}
	return temp;
}

int part1(std::map<std::string,std::vector<std::string>>& adj) {
	int total = 0;
	std::queue<std::string> q({std::move(std::string("you"))});
	while(!q.empty()) {
		std::string u(std::move(q.front())); q.pop();
		if(u == "out") {
			total++; continue;
		}
		for(const auto& v: adj[u]) {
			q.push(v);
		}
	}
	return total;
}

std::vector<std::string> topsort(std::map<std::string,std::vector<std::string>>& adj) {
	std::map<std::string,int> indegree;
	for(const auto& [u, edges]: adj) {
		if(indegree.find(u) == indegree.end()) {
			indegree[u] = 0;
		}
		for(const auto& v: edges) {
			indegree[v]++;
		}
	}
	std::queue<std::string> q;
	for(const auto& [u,in]: indegree) {
		if(in == 0) {
			q.push(u);
		}
	}
	std::vector<std::string> topord;
	while(!q.empty()) {
		std::string u(std::move(q.front())); q.pop();
		topord.push_back(u);
		for(const auto& v: adj[u]) {
			if(v == "out") {
				continue;
			}
			indegree[v]--;
			if(indegree[v] == 0) {
				q.push(v);
			}
		}
	}
	topord.push_back(std::string("out"));
	return topord;
}

long long paths(std::map<std::string,std::vector<std::string>>& adj, std::vector<std::string>& topord, std::string&& src, std::string&& dst)
{
	std::map<std::string,long long> dp;
	dp[src] = 1;
	auto begin = std::find(topord.begin(), topord.end(), src); 
	auto end = std::find(topord.begin(), topord.end(), dst);
	for(auto it = begin; it < end; ++it) {
		std::string u(*it);
		for(const auto& v: adj[u]) {
			dp[v] += dp[u];
		}
	}
	return dp[dst];
}
long long part2(std::map<std::string,std::vector<std::string>>& adj) {
	std::vector<std::string> topord(std::move(topsort(adj)));

	auto fft_pos = std::find(topord.begin(), topord.end(), std::string("fft"));
	auto dac_pos = std::find(topord.begin(), topord.end(), std::string("dac"));

	std::string first("svr");
	std::string second = (dac_pos < fft_pos) ? "dac" : "fft";
	std::string third = (second == "dac") ? "fft" : "dac";
	std::string last("out");

	long long x = paths(adj,topord,std::forward<std::string>(first),std::forward<std::string>(second));
	long long y = paths(adj,topord,std::forward<std::string>(second),std::forward<std::string>(third));
	long long z = paths(adj,topord,std::forward<std::string>(third),std::forward<std::string>(last));
	return x * y * z;
}

int main() {
	std::fstream input("data/day11.txt");
	std::map<std::string,std::vector<std::string>> adj;
	std::string s;
	
	int i = 0;
	while(std::getline(input,s)) {
		std::vector<std::string> line(std::move(split(std::forward<std::string>(s),':')));
		std::string u(std::move(line[0]));
		std::vector<std::string> edges(std::move(split(std::forward<std::string>(line[1].substr(1)),' ')));
		adj[u] = edges;
	}
	
	std::cout << "Part 1: " << part1(adj) << "\n";
	std::cout << "Part 2: " << part2(adj) << "\n";
	return 0;
}

File: cpp/day11.py
from heapq import heappush, heappop
from collections import defaultdict

def part1(adj):
	q = []
	heappush(q, "you")
	paths = 0
	while q:
		u = heappop(q)
		if u == "out":
			paths += 1
			continue
		for v in adj[u]:
			heappush(q, v)
	return paths

def topsort(adj):
	indegree = defaultdict(lambda: 0)
	for u, edges in adj.items():
		for v in edges:
			indegree[v] += 1
	
	q = []
	for v in adj.keys():
		if indegree[v] == 0:
			heappush(q, v)
	
	topord = []
	while q:
		u = heappop(q)
		topord.append(u)
		for v in adj[u]:
			if v == "out":
				continue
			indegree[v] -= 1
			if indegree[v] == 0:
				heappush(q, v)
	
	topord.append("out")
	return topord

def paths(adj,topord,src,dst):
	dp = defaultdict(lambda: 0)
	dp[src] = 1; i = topord.index(src); j = topord.index(dst)
	
	for u in topord[i:j+1]:
		if dp[u] > 0:
			if u == "out":
				break
			for v in adj[u]:
				dp[v] += dp[u]
	
	return dp[dst]

def part2(adj):
	topord = topsort(adj)
	x = paths(adj,topord,"svr","fft")
	y = paths(adj,topord,"fft","dac")
	z = paths(adj,topord,"dac","out")
	return x*y*z

def main():
	adj = {}

	with open("data/day11.txt", "r") as f:
		lines = f.read().splitlines()
		for i, line in enumerate(lines):
			u = line.split(":")[0]
			edges = line.split(":")[1].split(" ")[1:]
			adj[u] = edges

	print(part1(adj))
	print(part2(adj))
if __name__ == "__main__":
	main()


